tempesta_fw/t/unit/tfw_fuzzer.c

```diff
@@ -0,0 +1,563 @@
+
+static struct {
+	int i;
+	int max;
+	char *a_val;
+	char *a_inval;
+	int singular; /* only for headers; 0 - nonsingular, 1 - singular */
+} gen_vector[] = {
+	/* METHOD */
+	{0, sizeof(methods) / sizeof(struct fuzz_msg) - 1, "", NULL},
+	/* SPACES */
+	{1, 6, "", NULL},
+	/* HTTP_VER */
+	{0, sizeof(versions) / sizeof(struct fuzz_msg) - 1, "", NULL},
+	/* RESP_CODE */
+	{0, sizeof(resp_code) / sizeof(struct fuzz_msg) - 1, "", NULL},
+	/* URI_PATH_START */
+	{0, sizeof(uri_path_start) / sizeof(struct fuzz_msg) - 1, "", NULL},
+	/* URI_PATH_DEPTH */
+	{1, 3, "", NULL},
+	/* URI_FILE */
+	{0, sizeof(uri_file) / sizeof(struct fuzz_msg) - 1, A_URI, NULL},
+	/* CONNECTION */
+	{0, sizeof(conn_val) / sizeof(struct fuzz_msg) - 1, "", NULL, 1},
+	/* USER_AGENT*/
+	{0, sizeof(ua_val) / sizeof(struct fuzz_msg) - 1, A_USER_AGENT,
+		A_USER_AGENT_INVAL, 1},
+	/* HOST */
+	{0, sizeof(hosts) / sizeof(struct fuzz_msg) - 1, A_HOST,
+		A_HOST_INVAL, 1},
+	/* X_FORWARDED_FOR */
+	{0, sizeof(hosts) / sizeof(struct fuzz_msg) - 1, A_X_FORWARDED_FOR,
+		A_X_FORWARDED_FOR_INVAL, 0},
+	/* CONTENT_TYPE */
+	{0, sizeof(content_type) / sizeof(struct fuzz_msg) - 1, "", NULL, 1},
+	/* CONTENT_LENGTH */
+	{0, sizeof(content_length) / sizeof(struct fuzz_msg) - 1, "0123456789",
+		NULL, 1},
+	/* TRANSFER_ENCODING */
+	{0, sizeof(transfer_encoding) / sizeof(struct fuzz_msg) - 1, "", NULL, 0},
+	/* TRANSFER_ENCODING_NUM */
+	{0, 5, "", NULL},
+	/* DUPLICATES */
+	{0, 3, "", NULL},
+	/* BODY_SIZE */
+	{0, 10, "", NULL}
+};
+
+enum {
+	METHOD,
+	SPACES,
+	HTTP_VER,
+	RESP_CODE,
+	URI_PATH_START,
+	URI_PATH_DEPTH,
+	URI_FILE,
+	CONNECTION,
+	USER_AGENT,
+	HOST,
+	X_FORWARDED_FOR,
+	CONTENT_TYPE,
+	CONTENT_LENGTH,
+	TRANSFER_ENCODING,
+	TRANSFER_ENCODING_NUM,
+	DUPLICATES,
+	BODY_SIZE,
+	N_FIELDS
+};
+
+static int gen_vector_move(int i)
+{
+	if (gen_vector[i].i++ == gen_vector[i].max) {
+		if (i == 0)
+			return FUZZ_END;
+		gen_vector[i].i = 0;
+		if (gen_vector_move(i - 1) == FUZZ_END)
+			return FUZZ_END;
+	}
+
+	return FUZZ_VALID;
+}
+
+static char *addch(char *s1, char ch)
+{
+	size_t len;
+
+	if (!ch)
+		return s1;
+	len = strlen(s1);
+	s1[len] = ch;
+	s1[len + 1] = '\0';
+
+	return s1;
+}
+
+static char *add_string(char *s1, int i, char *seed)
+{
+	int j;
+
+	if (seed == NULL)
+		seed = "\03\r\023\07";
+	if (!strlen(seed))
+		return s1;
+
+	for (j = 0; j < i; ++j)
+		addch(s1, seed[((j + 333) ^ seed[j]) % sizeof(seed)]);
+
+	return s1;
+}
+
+static char *add_spaces(char *s1)
+{
+	int i;
+
+	for (i = 0; i < gen_vector[SPACES].i; ++i) {
+		printk("%s:i:%d\n", __func__, i);
+		if (i % 2)
+			addch(s1, '\t');
+		else if (i % 3)
+			addch(s1, '\r');
+		else
+			addch(s1, ' ');
+	}
+
+	return s1;
+}
+
+static int add_method(char *s1)
+{
+	struct fuzz_msg r = methods[gen_vector[METHOD].i];
+
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[METHOD].i,
```

Comment:
gen_vector[METHOD].i = 0, so do we generate empty method here and return FUZZ_VALID?